lib/fdtable.c

```diff
@@ -0,0 +1,172 @@
+/*
+ * Copyright (c) 2018 Linaro Limited
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+/**
+ * @file
+ * @brief File descriptor table
+ *
+ * This file provides generic file descriptor table implementation, suitable
+ * for any I/O object implementing POSIX I/O semantics (i.e. read/write +
+ * aux operations).
+ */
+
+#include <errno.h>
+#include <kernel.h>
+#include <misc/fdtable.h>
+
+struct fd_entry {
+	void *obj;
+	const struct fd_op_vtable *vtable;
+};
+
+static struct fd_entry fdtable[CONFIG_POSIX_MAX_FDS];
+
+static int _find_fd_entry(void)
+{
+	int fd;
+
+	for (fd = 0; fd < ARRAY_SIZE(fdtable); fd++) {
+		if (fdtable[fd].obj == NULL) {
+			return fd;
+		}
+	}
+
+	errno = ENFILE;
+	return -1;
+}
+
+static int _check_fd(int fd)
+{
+	if (fd < 0 || fd >= ARRAY_SIZE(fdtable) || fdtable[fd].obj == NULL) {
+		errno = EBADF;
+		return -1;
+	}
+
+	return 0;
+}
+
+void *z_get_fd_obj(int fd, const struct fd_op_vtable *vtable, int err)
+{
+	struct fd_entry *fd_entry;
+
+	if (_check_fd(fd) < 0) {
+		return NULL;
+	}
+
+	fd_entry = &fdtable[fd];
+
+	if (vtable != NULL && fd_entry->vtable != vtable) {
+		errno = err;
+		return NULL;
+	}
+
+	return fd_entry->obj;
+}
+
+int z_reserve_fd(void)
+{
+	int fd;
+	unsigned int key;
+
+	key = irq_lock();
```

Comment:
> Is the intention that z_reserve_fd() might be called from an ISR?

Nope, it's just a spinlock, a lightweight critical section, to protect just a few instructions. Well, maybe it might make sense to use mutex, as after all there's a loop, kinda O(n) operation...